EasyPost.Tests/ServicesTests/PartnerServiceTest.cs

```diff
@@ -0,0 +1,234 @@
+                    new TestUtils.MockRequestMatchRules(Method.Get, @"^v2\/partners\/stripe_public_key$"),
+                    new TestUtils.MockRequestResponseInfo(HttpStatusCode.OK, content: "{\"public_key\":\"pk_test_12345\"}")
+                ),
+                new(
+                    new TestUtils.MockRequestMatchRules(Method.Post, @"^https://api.stripe.com/v1/tokens$"),
+                    new TestUtils.MockRequestResponseInfo(HttpStatusCode.NotFound, content: "{}")
+                )
+            });
+
+            CreditCard card = new CreditCard(Fixtures.CreditCardDetails);
+
+            await Assert.ThrowsAsync<ExternalApiError>(async () => await Client.Partner.AddCreditCardToUser(ReferralUserKey, card.Number, card.ExpirationMonth, card.ExpirationYear, card.Cvc, PaymentMethod.Priority.Primary));
+        }
+
+        [Fact]
+        [Testing.Exception]
+        public async Task TestAddCreditCardToUserStripeBadResponse()
+        {
+            UseMockClient(new List<TestUtils.MockRequest>
+            {
+                new(
+                    new TestUtils.MockRequestMatchRules(Method.Get, @"^v2\/partners\/stripe_public_key$"),
+                    new TestUtils.MockRequestResponseInfo(HttpStatusCode.OK, content: "{\"public_key\":\"pk_test_12345\"}")
+                ),
+                new(
+                    new TestUtils.MockRequestMatchRules(Method.Post, @"^https://api.stripe.com/v1/tokens$"),
+                    new TestUtils.MockRequestResponseInfo(HttpStatusCode.OK, content: "{\"not_id\":\"random\"}")
+                )
+            });
+
+            CreditCard card = new CreditCard(Fixtures.CreditCardDetails);
+
+            await Assert.ThrowsAsync<ExternalApiError>(async () => await Client.Partner.AddCreditCardToUser(ReferralUserKey, card.Number, card.ExpirationMonth, card.ExpirationYear, card.Cvc, PaymentMethod.Priority.Primary));
+        }
+
+        [Fact]
+        [CrudOperations.Update]
+        [Testing.Function]
+        public async Task TestUpdateReferralEmail()
+        {
+            UseVCR("update_referral_email");
+
+            ReferralCustomer referralCustomer = await CreateReferral();
+
+            Exception? possibleException = await Record.ExceptionAsync(async () => await Client.Partner.UpdateReferralEmail(referralCustomer.Id, "email@example.com"));
+
+            Assert.Null(possibleException);
+        }
+
+        #endregion
+
+        #endregion
+
+        private static string ReferralUserKey => TestUtils.GetApiKey(TestUtils.ApiKey.Referral);
+
+        private async Task<ReferralCustomer> CreateReferral() => await Client.Partner.CreateReferral(Fixtures.ReferralUser);
+
+        private class CreditCard
+        {
+            internal readonly string Number;
+
+            internal readonly int ExpirationMonth;
+
+            internal readonly int ExpirationYear;
+
+            internal readonly string Cvc;
+
+            internal CreditCard(IReadOnlyDictionary<string, object> details)
```

Comment:
Question: Should this not already be a fixture we can use instead of needing to recreate it?

Reply:
The `CreditCardDetails` fixture is a dictionary. There was a lot of repetitive casting going on that I wanted to consolidate.